sdk/Trace/Span.php

```diff
@@ -59,383 +81,374 @@ public static function fromContext(Context $context): API\Span
+                ' at %s%s%s(%s%s%s)',
+                count($trace) && array_key_exists('class', $trace[0]) ? str_replace('\\', '.', $trace[0]['class']) : '',
+                count($trace) && array_key_exists('class', $trace[0]) && array_key_exists('function', $trace[0]) ? '.' : '',
+                count($trace) && array_key_exists('function', $trace[0]) ? str_replace('\\', '.', $trace[0]['function']) : 'main',
+                $line === null ? $file : basename($file),
+                $line === null ? '' : ':',
+                $line ?? ''
+            );
+            $seen[] = "$file:$line";
+            if (!count($trace)) {
+                break;
+            }
+            $file = array_key_exists('file', $trace[0]) ? $trace[0]['file'] : 'Unknown Source';
+            $line = array_key_exists('file', $trace[0]) && array_key_exists('line', $trace[0]) && $trace[0]['line'] ? $trace[0]['line'] : null;
+            array_shift($trace);
+        }
+        $result = implode("\n", $result);
+        if ($prev) {
+            $result  .= "\n" . self::formatStackTrace($prev, $seen);
+        }
 
-        // todo: set these to null until needed
-        $this->attributes = $attributes ?? new Attributes();
-        $this->events = new Events();
-        $this->links = $links ?? new Links();
+        return $result;
     }
 
-    /**
-     * @internal
-     */
-    public function setInstrumentationLibrary(InstrumentationLibrary $instrumentationLibrary)
+    /** @inheritDoc */
+    public static function wrap(API\SpanContext $spanContext): API\Span
     {
-        $this->instrumentationLibrary = $instrumentationLibrary;
+        return new NonRecordingSpan($spanContext);
     }
 
-    /**
-     * @internal
-     */
-    public function getInstrumentationLibrary(): InstrumentationLibrary
-    {
-        return $this->instrumentationLibrary;
-    }
+    // TODO: Add a SpanLimits object to the constructor for configuration options.
 
-    public function getResource(): ResourceInfo
-    {
-        return clone $this->resource;
-    }
+    /** @readonly */
+    private API\SpanContext $context;
 
-    public function getParent(): ?API\SpanContext
-    {
-        // todo: Spec says a parent is a Span, SpanContext, or null -> should we implement this here?
-        return $this->parentSpanContext !== null ? clone $this->parentSpanContext : null;
-    }
+    /** @readonly */
+    private API\SpanContext $parentSpanContext;
+
+    /** @readonly */
+    private SpanProcessor $spanProcessor;
 
     /**
-     * @param string $code
-     * @param string|null $description
-     * @return Span
+     * @readonly
+     *
+     * @todo: Java just has this as list<API\Event>, could we just do that?
      */
-    public function setSpanStatus(string $code, ?string $description = null): API\Span
-    {
-        if ($this->isRecording()) {
-            $this->spanStatus->setStatus($code, $description);
-        }
+    private API\Events $events;
```

Comment:
I'm skeptical if we really need these `API\Events` and `API\Links` objects versus just using arrays, or `ArrayObject`.

Reply:
You think they might just be vestegial?

Reply:
Or at least unnecessary. It's just trying to represent something that can be traversed, counted, and have things added to it. I.e. essentially a https://www.php.net/manual/en/class.arrayobject.php. Not really providing any meaningful abstraction/benefit.

Reply:
Moreover an additional layer of abstraction gets `psalm` crazy, it can't infer a correct type of `$event` in `foreach ($events as $event)` loop.